### PR TITLE
Optimize away Rectangle that serve no purpose

### PR DESCRIPTION
**Pull request by ogoffart**
Rectangles which do not draw anything and have no x or y don't need to be in
the item tree, we can just remove them.

### CODE REVIEW

File: sixtyfps_compiler/layout.rs [hunk 1]
```diff
@@ -454,3 +454,34 @@ pub fn layout_info_type() -> Type {
         node: None,
     }
 }
+
+/// Get the implicit layout info of a particular element
+pub fn implicit_layout_info_call(elem: &ElementRc, orientation: Orientation) -> Expression {
+    if matches!(&elem.borrow().base_type ,
+                    Type::Builtin(base_type) if base_type.name == "Rectangle")
```
Comment (tronical) [inline sixtyfps_compiler/layout.rs:461]: We have this pattern in a few places, might become worth a helper function in Element

**Review by tronical — APPROVED** [review]
Nice :)

**Review by tronical — APPROVED** [review]

File: sixtyfps_compiler/passes/optimize_useless_rectangles.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+//!
+//! Rectangles which do not draw anything and have no x or y don't need to be in
+//! the item tree, we can just remove them.
+
+use crate::{langtype::Type, object_tree::*};
+use std::rc::Rc;
+
+pub fn optimize_useless_rectangles(root_component: &Rc<Component>) {
+    let mut item_count = 0;
+    let mut optim_count = 0;
```
Comment (tronical) [inline sixtyfps_compiler/passes/optimize_useless_rectangles.rs:20]: I suppose that this is unused?:)

File: sixtyfps_compiler/passes/optimize_useless_rectangles.rs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    if e.is_flickable_viewport || e.child_of_layout {
+        return false;
+    };
+
+    let base_type = match &e.base_type {
+        Type::Builtin(base_type) if base_type.name == "Rectangle" => base_type,
+        _ => return false,
+    };
+
+    // Check that no Rectangle property other than height and width are set
```
Comment (tronical) [inline sixtyfps_compiler/passes/optimize_useless_rectangles.rs:54]: What happens for example if a rectangle sets just width and height and a child binds to the parent’s width/height?

Comment (ogoffart) [reply sixtyfps_compiler/passes/optimize_useless_rectangles.rs:54]: in that case, the height and width properties will be materialized by the move_declarations pass (via move_declarations::simplify_optimized_items )

Comment (tronical) [reply sixtyfps_compiler/passes/optimize_useless_rectangles.rs:54]: Ah yes, that makes sense

**Pull request merged**